### PR TITLE
feat: Add Copy command parsing in QueryMessage and basic psql e2e test

### PR DESCRIPTION
**Pull request by Vizerai**

### CODE REVIEW

File: src/main/java/com/google/cloud/spanner/pgadapter/wireprotocol/ControlMessage.java
Comment (tinaspark) [inline src/main/java/com/google/cloud/spanner/pgadapter/wireprotocol/ControlMessage.java:61]: should we handle CopyFail here as well?

File: src/main/java/com/google/cloud/spanner/pgadapter/wireprotocol/ControlMessage.java
Comment (tinaspark) [inline src/main/java/com/google/cloud/spanner/pgadapter/wireprotocol/ControlMessage.java]: are we allowed to receive CopyDone/CopyData/CopyFail if we're not in COPY_IN mode?

File: src/main/java/com/google/cloud/spanner/pgadapter/wireprotocol/CopyDoneMessage.java
Comment (olavloite) [inline src/main/java/com/google/cloud/spanner/pgadapter/wireprotocol/CopyDoneMessage.java:56]: Sorry, I didn't notice this before, but don't we need to return something like a `ReadyResponse` after this? Or will the exception that is thrown ensure that the `ConnectionHandler` returns a response?

File: src/main/java/com/google/cloud/spanner/pgadapter/wireprotocol/CopyDoneMessage.java
Comment (olavloite) [inline src/main/java/com/google/cloud/spanner/pgadapter/wireprotocol/CopyDoneMessage.java:64]: I think that this call to `setStatus(..)` and also the above one are causing a build error, because `ConnectionStatus` is not imported in this file. Would you mind taking a look at that?

**Review by olavloite — COMMENTED** [review]
The build fails because of a compile error, I think because `ConnectionStatus` is not imported by `CopyDoneMessage`.

File: src/main/java/com/google/cloud/spanner/pgadapter/wireprotocol/CopyDoneMessage.java
Comment (Vizerai) [reply src/main/java/com/google/cloud/spanner/pgadapter/wireprotocol/CopyDoneMessage.java:56]: The ConnectionHandler will return a ready response of IDLE on handling an error.

File: src/main/java/com/google/cloud/spanner/pgadapter/wireprotocol/CopyDoneMessage.java
Comment (Vizerai) [reply src/main/java/com/google/cloud/spanner/pgadapter/wireprotocol/CopyDoneMessage.java:64]: Fixed now.  Originally this PR was supposed to require changes from https://github.com/GoogleCloudPlatform/pgadapter/pull/40

File: src/main/java/com/google/cloud/spanner/pgadapter/wireprotocol/ControlMessage.java
Comment (Vizerai) [reply src/main/java/com/google/cloud/spanner/pgadapter/wireprotocol/ControlMessage.java]: No, it should not.

File: src/main/java/com/google/cloud/spanner/pgadapter/wireprotocol/ControlMessage.java
Comment (Vizerai) [reply src/main/java/com/google/cloud/spanner/pgadapter/wireprotocol/ControlMessage.java:61]: Done.

**Comment by Vizerai** [comment]
The e2e failures seem to be due to a change on the backend.  There a new pg_catalog entries returned from the information schema.

**Comment by olavloite** [comment]
The current e2e test failure should be fixed once https://github.com/GoogleCloudPlatform/pgadapter/pull/40 has been merged (it's caused by the "?" that is fixed in #40)

**Review by tinaspark — COMMENTED** [review]
LGTM once the test failures are fixed by #40

**Review by olavloite — APPROVED** [review]

**Comment by olavloite** [comment]
@Vizerai I was wondering why the e2e-tests were failing, so I had a look at this PR and ended up doing a couple of small changes. It might be good if you go through those to check that you are OK with them before merging.

**Comment by Vizerai** [comment]
> @Vizerai I was wondering why the e2e-tests were failing, so I had a look at this PR and ended up doing a couple of small changes. It might be good if you go through those to check that you are OK with them before merging.

The changes look good.  Thanks for the update.

File: src/main/java/com/google/cloud/spanner/pgadapter/statements/CopyStatement.java
Comment (tinaspark) [inline src/main/java/com/google/cloud/spanner/pgadapter/statements/CopyStatement.java:175]: why do we need to issue this as a GoogleSQL query?

Comment (Vizerai) [reply src/main/java/com/google/cloud/spanner/pgadapter/statements/CopyStatement.java:175]: The backend no longer returns these as GoogleSQL types for this query. It instead returns the PG types now. We need the GoogleSQL types to fill out the mutations.

This was just recently changed:
https://buganizer.corp.google.com/issues/198312860

Comment (tinaspark) [reply src/main/java/com/google/cloud/spanner/pgadapter/statements/CopyStatement.java:175]: I see. Can you add a comment?

// The mutation API requires GoogleSQL type names instead of PostgreSQL type names for the table column types
// Issue the information_schema table query as a GoogleSQL query so that it will return GoogleSQL type names.

**Review by tinaspark — APPROVED** [review]

File: src/main/java/com/google/cloud/spanner/pgadapter/statements/CopyStatement.java
Comment (Vizerai) [reply src/main/java/com/google/cloud/spanner/pgadapter/statements/CopyStatement.java:175]: Done.

**Pull request merged**
